bundles/org.openhab.binding.pentair/src/main/java/org/openhab/binding/pentair/internal/handler/PentairControllerHandler.java

```diff
@@ -0,0 +1,1186 @@
+            }
+        }
+
+        if (fullChannelUID == null || channelID.equals(CONTROLLER_SCHEDULESTRING)) {
+            updateChannel(schedule.getGroupID(), CONTROLLER_SCHEDULESTRING, schedule.toString());
+        }
+
+        if (fullChannelUID == null || channelID.equals(CONTROLLER_SCHEDULETYPE)) {
+            logger.debug("groupID: {}, scheduleType: {}", schedule.getGroupID(), schedule.getScheduleTypeStr());
+            updateChannel(schedule.getGroupID(), CONTROLLER_SCHEDULETYPE, schedule.getScheduleTypeStr());
+        }
+
+        if (fullChannelUID == null || channelID.equals(CONTROLLER_SCHEDULECIRCUIT)) {
+            updateChannel(schedule.getGroupID(), CONTROLLER_SCHEDULECIRCUIT, schedule.circuit);
+        }
+
+        if (fullChannelUID == null || channelID.equals(CONTROLLER_SCHEDULESTART)) {
+            updateChannel(schedule.getGroupID(), CONTROLLER_SCHEDULESTART, schedule.start);
+        }
+
+        if (fullChannelUID == null || channelID.equals(CONTROLLER_SCHEDULEEND)) {
+            updateChannel(schedule.getGroupID(), CONTROLLER_SCHEDULEEND, schedule.end);
+        }
+
+        if (fullChannelUID == null || channelID.equals(CONTROLLER_SCHEDULEDAYS)) {
+            updateChannel(schedule.getGroupID(), CONTROLLER_SCHEDULEDAYS, schedule.getDays());
+        }
+
+        logger.debug(
+                "Controller Schedule - ID: {}, Type: {}, Circuit: {}, Start Time: {}:{}, End Time: {}:{}, Days: {}",
+                schedule.id, schedule.type, schedule.circuit, schedule.start / 60, schedule.start % 60,
+                schedule.end / 60, schedule.end % 60, schedule.getDays());
+    }
+
+    public void refreshControllerStatusChannels(@Nullable ChannelUID fullChannelUID,
+            PentairControllerStatus controllerStatus) {
+        String groupID = "";
+        String channelID = "";
+
+        if (fullChannelUID != null) {
+            groupID = Objects.requireNonNull(fullChannelUID.getGroupId());
+            channelID = fullChannelUID.getIdWithoutGroup();
+        }
+
+        // update circuit states
+        if (channelID.equals(CONTROLLER_CIRCUITSWITCH)) {
+            PentairControllerCircuit circuit;
+
+            circuit = getCircuitByGroupID(groupID);
+            if (circuit == null) {
+                return;
+            }
+
+            int circuitNum = circuit.id - 1;
+            updateChannel(groupID, CONTROLLER_CIRCUITSWITCH, controllerStatus.circuits[circuitNum]);
+        } else if (fullChannelUID == null) {
+            for (int i = 0; i < NUM_CIRCUITS; i++) {
+                String circuitGroupID;
+
+                circuitGroupID = PentairControllerCircuit.getCircuitGroupIDByNumber(i + 1);
+
+                if (circuitGroupID != null) {
+                    updateChannel(circuitGroupID, CONTROLLER_CIRCUITSWITCH, controllerStatus.circuits[i]);
+                }
+            }
+        }
+
+        if (fullChannelUID == null || channelID.equals(CONTROLLER_UOM)) {
```

Comment:
Please save `fullChannelUID == null` in a boolean local variable so you aren't constantly calculating it.